service/matching/task_queue_partition_manager.go

```diff
@@ -197,6 +250,21 @@ func (pm *taskQueuePartitionManagerImpl) AddTask(
 ) (buildId string, syncMatched bool, err error) {
 	var spoolQueue, syncMatchQueue physicalTaskQueueManager
 	directive := params.taskInfo.GetVersionDirective()
+
+	if pm.Partition().IsRoot() && pm.config.AutoEnable && pm.fairnessState == persistencespb.TaskQueueTypeUserData_FAIRNESS_STATE_UNSPECIFIED {
+		if params.taskInfo.Priority != nil && (params.taskInfo.Priority.FairnessKey != "" || params.taskInfo.Priority.PriorityKey != int32(0)) {
+			updateFn := func(old *persistencespb.TaskQueueUserData) (*persistencespb.TaskQueueUserData, bool, error) {
+				data := common.CloneProto(old)
+				perType := data.GetPerType()[int32(pm.Partition().TaskType())]
+				perType.FairnessState = persistencespb.TaskQueueTypeUserData_FAIRNESS_STATE_V2
+				return data, true, nil
+			}
+			_, err := pm.userDataManager.UpdateUserData(ctx, UserDataUpdateOptions{Source: "Matching auto enable"}, updateFn)
```

Comment:
unfortunately, we decided to store user data for all task queue types on the root of the _workflow_ task queue. so this local call can only be done on the workflow tq. for activity, we need to do an rpc to the worker tq.